hydra-node/test/Hydra/HeadLogicSpec.hs

```diff
@@ -562,32 +573,39 @@ data StepState tx = StepState
   }
 
 -- | Asserts that the update function will update the state (return a NewState) for this Event
-step :: (MonadState (StepState tx) m, HasCallStack, IsChainState tx) => Event tx -> m (HeadState tx)
+step ::
+  (MonadState (StepState tx) m, HasCallStack, IsChainState tx, MonadCatch m) =>
+  Event tx ->
+  m (HeadState tx)
 step event = do
-  StepState{ headState, env, ledger} <- get
+  StepState{headState, env, ledger} <- get
   headState' <- assertNewState $ update env ledger headState event
-  put StepState{ env, ledger, headState = headState' }
+  put StepState{env, ledger, headState = headState'}
   pure headState'
 
 assertNewState ::
-  (HasCallStack, IsChainState tx, Applicative m) =>
+  (HasCallStack, IsChainState tx, MonadCatch m) =>
   Outcome tx ->
   m (HeadState tx)
 assertNewState = \case
-  NewState st _ -> pure st
-  OnlyEffects effects -> Hydra.Prelude.error $ "Unexpected 'OnlyEffects' outcome: " <> show effects
-  Error e -> Hydra.Prelude.error $ "Unexpected 'Error' outcome: " <> show e
-  Wait r -> Hydra.Prelude.error $ "Unexpected 'Wait' outcome with reason: " <> show r
+  NewState st -> pure st
+  Effects effects -> failure $ "Unexpected 'OnlyEffects' outcome: " <> show effects
+  Error e -> failure $ "Unexpected 'Error' outcome: " <> show e
+  Wait r -> failure $ "Unexpected 'Wait' outcome with reason: " <> show r
+  Combined l r -> assertNewState l `orElse` assertNewState r
 
-assertOnlyEffects ::
+assertEffects ::
   (HasCallStack, IsChainState tx) =>
   Outcome tx ->
   IO ()
-assertOnlyEffects = \case
-  NewState st _ -> failure $ "Unexpected 'NewState' outcome: " <> show st
-  OnlyEffects _ -> pure ()
+assertEffects = \case
```

Comment:
Should factorize by calling hasEffectSatisfying with the predicate `True`

Reply:
assertEffects and hasEffectSatisfying are testing two different things, the former tests there are only effects without caring about the outcome, whereas the later accepts effects and other things, so in summary they have different semantics.

by @abailly-iohk